### ISSUE TITLE
Flux Lora training makes no tangible difference

### ISSUE THREAD

**Issue by Tophness** [kohya-ss/sd-scripts#1464]
Not sure what I'm doing wrong, but I trained this concept lora on the same dataset I've used to train SDXL models (239+87 images, 2669 reg images) using the recommended settings and it makes virtually no difference to the outputs in comfy or using flux_minimal_inference.py. A leaf on a tree might move an inch and that's about it. Same person in the same pose in the same clothes, everything else in the image is the same.
Anyone else getting this? Here's what I used:

`accelerate launch --mixed_precision bf16 --num_cpu_threads_per_process 1 flux_train_network.py --pretrained_model_name_or_path flux1-dev.safetensors --clip_l clip_l.safetensors --t5xxl t5xxl_fp16.safetensors --ae ae.safetensors --cache_latents_to_disk --save_model_as safetensors --sdpa --persistent_data_loader_workers --max_data_loader_n_workers 2 --seed 42 --gradient_checkpointing --mixed_precision bf16 --save_precision bf16 --network_module networks.lora_flux --network_dim 4 --learning_rate 1e-4 --network_train_unet_only --cache_text_encoder_outputs --cache_text_encoder_outputs_to_disk --fp8_base --max_train_epochs 4 --save_every_n_epochs 1 --dataset_config model.toml --output_dir=outputs --output_name model --timestep_sampling sigmoid --model_prediction_type raw --guidance_scale 1.0 --loss_type l2 --optimizer_type adafactor --optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False"`

**Comment by sipie800** [comment]
Same here. I trained a real person lora with schnell. The trained weight doesn't hear anything at inferencing. Didn't figure out what.
Is the provided default parameters proper of flux_train_network? Or it's just far from a feasible config ?

**Comment by sipie800** [comment]
The default parameter is not proper. Need to lift the learning, I increase epoch from default 4 to 50 for real person with 50 images. Got pretty good result. It's better than SDXL.

**Comment by bghira** [comment]
rank 16 for me had better skin details than 128

**Comment by envy-ai** [comment]
@Tophness I'm having the same problem...  Check your training output.  I'm getting the average loss showing up as NaN, and I'm suspicious that probably has something to do with this?

```
FLUX: Gradient checkpointing enabled.
prepare optimizer, data loader etc.
                    INFO     use Prodigy optimizer | {}                                               train_util.py:4491
enable fp8 training.
running training / 学習開始
  num train images * repeats / 学習画像の数×繰り返し回数: 105
  num reg images / 正則化画像の数: 0
  num batches per epoch / 1epochのバッチ数: 105
  num epochs / epoch数: 16
  batch size per device / バッチサイズ: 1
  gradient accumulation steps / 勾配を合計するステップ数 = 1
  total optimization steps / 学習ステップ数: 1600
steps:   0%|                                                                                  | 0/1600 [00:00<?, ?it/s]2024-08-16 10:58:26 INFO     unet dtype: torch.float8_e4m3fn, device: cuda:0                       train_network.py:1004
                    INFO     text_encoder dtype: torch.float8_e4m3fn, device: cpu                  train_network.py:1006
                    INFO     text_encoder dtype: torch.float8_e4m3fn, device: cpu                  train_network.py:1006

epoch 1/16
                    INFO     epoch is incremented. current_epoch: 0, epoch: 1                          train_util.py:668
C:\Users\elbar\anaconda3\envs\kohya_flux\lib\site-packages\torch\utils\checkpoint.py:1399: FutureWarning: `torch.cpu.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cpu', args...)` instead.
  with device_autocast_ctx, torch.cpu.amp.autocast(**cpu_autocast_kwargs), recompute_context:  # type: ignore[attr-defined]
steps:   7%|███▋                                                    | 105/1600 [04:14<1:00:20,  2.42s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000001.safetensors

epoch 2/16
2024-08-16 11:02:41 INFO     epoch is incremented. current_epoch: 1, epoch: 2                          train_util.py:668
steps:  13%|███████▌                                                  | 210/1600 [08:27<56:00,  2.42s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000002.safetensors

epoch 3/16
2024-08-16 11:06:54 INFO     epoch is incremented. current_epoch: 2, epoch: 3                          train_util.py:668
steps:  20%|███████████▍                                              | 315/1600 [12:40<51:41,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000003.safetensors

epoch 4/16
2024-08-16 11:11:07 INFO     epoch is incremented. current_epoch: 3, epoch: 4                          train_util.py:668
steps:  26%|███████████████▏                                          | 420/1600 [16:53<47:26,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000004.safetensors

epoch 5/16
2024-08-16 11:15:20 INFO     epoch is incremented. current_epoch: 4, epoch: 5                          train_util.py:668
steps:  33%|███████████████████                                       | 525/1600 [21:06<43:13,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000005.safetensors

epoch 6/16
2024-08-16 11:19:33 INFO     epoch is incremented. current_epoch: 5, epoch: 6                          train_util.py:668
steps:  39%|██████████████████████▊                                   | 630/1600 [25:19<38:59,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000006.safetensors

epoch 7/16
2024-08-16 11:23:46 INFO     epoch is incremented. current_epoch: 6, epoch: 7                          train_util.py:668
steps:  46%|██████████████████████████▋                               | 735/1600 [29:33<34:46,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000007.safetensors

epoch 8/16
2024-08-16 11:27:59 INFO     epoch is incremented. current_epoch: 7, epoch: 8                          train_util.py:668
steps:  52%|██████████████████████████████▍                           | 840/1600 [33:46<30:33,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000008.safetensors

epoch 9/16
2024-08-16 11:32:12 INFO     epoch is incremented. current_epoch: 8, epoch: 9                          train_util.py:668
steps:  59%|██████████████████████████████████▎                       | 945/1600 [38:00<26:20,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000009.safetensors

epoch 10/16
2024-08-16 11:36:26 INFO     epoch is incremented. current_epoch: 9, epoch: 10                         train_util.py:668
steps:  66%|█████████████████████████████████████▍                   | 1050/1600 [42:13<22:07,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000010.safetensors

epoch 11/16
2024-08-16 11:40:40 INFO     epoch is incremented. current_epoch: 10, epoch: 11                        train_util.py:668
steps:  72%|█████████████████████████████████████████▏               | 1155/1600 [46:27<17:53,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000011.safetensors

epoch 12/16
2024-08-16 11:44:54 INFO     epoch is incremented. current_epoch: 11, epoch: 12                        train_util.py:668
steps:  79%|████████████████████████████████████████████▉            | 1260/1600 [50:40<13:40,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000012.safetensors

epoch 13/16
2024-08-16 11:49:07 INFO     epoch is incremented. current_epoch: 12, epoch: 13                        train_util.py:668
steps:  85%|████████████████████████████████████████████████▋        | 1365/1600 [54:54<09:27,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000013.safetensors

epoch 14/16
2024-08-16 11:53:21 INFO     epoch is incremented. current_epoch: 13, epoch: 14                        train_util.py:668
steps:  92%|████████████████████████████████████████████████████▎    | 1470/1600 [59:07<05:13,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000014.safetensors

epoch 15/16
2024-08-16 11:57:34 INFO     epoch is incremented. current_epoch: 14, epoch: 15                        train_util.py:668
steps:  98%|██████████████████████████████████████████████████████▏| 1575/1600 [1:03:21<01:00,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2-000015.safetensors

epoch 16/16
2024-08-16 12:01:47 INFO     epoch is incremented. current_epoch: 15, epoch: 16                        train_util.py:668
steps: 100%|███████████████████████████████████████████████████████| 1600/1600 [1:04:21<00:00,  2.41s/it, avr_loss=nan]
saving checkpoint: D:/ai/models/lora/flux_test2.safetensors
2024-08-16 12:02:48 INFO     model saved.                                                          train_network.py:1216
steps: 100%|███████████████████████████████████████████████████████| 1600/1600 [1:04:22<00:00,  2.41s/it, avr_loss=nan]
```

**Comment by envy-ai** [comment]
For the record, in my case, "no difference" means turning the LoRA strength up to 10 and literally nothing happens.

**Comment by Tophness** [comment]
> The default parameter is not proper. Need to lift the learning, I increase epoch from default 4 to 50 for real person with 50 images. Got pretty good result. It's better than SDXL. And the default lora dim is not proper IMO. I use 32 and 16. Flux is so good at details. To make good use of that, we may need higher rank in lora. And also, we may need higher quality with the images. If you use images with some artifects, the details (like hair texture...) may go bad than the origin flux. That would be a big mistake if you are using an excellent base model. To make good lora, we should take more care of the samples than with SDXL etc.

That makes sense. Never used anything as low as 4 epochs @ rank 4 before. Probably should update the advice in the readme as some point.

> rank 16 for me had better skin details than 128
Alright well I'll start there. I'm training a concept that applies to every person, so lower rank should be better anyway

> @Tophness I'm having the same problem... Check your training output. I'm getting the average loss showing up as NaN, and I'm suspicious that probably has something to do with this?

Nope, probably need to open a separate issue. I would check your dataset, the unet model and then go over your settings if you're getting NaN loss

**Comment by Tophness** [comment]
Just trained rank 128 for 3260 steps (equivalent of 4 epochs), still not learning the concept.
I'm going up to 30,000 so I'll update if this changes but I would've thought I'd see something by now.
Are we not supposed to use reg images with this model maybe?

**Comment by Tophness** [comment]
> What are the values ​​of dim (rank), alpha, learning rate, and type of optimizer for the latest training?
> 
> Edit: Please try training without regularization images. I am not sure how the reg images work on FLUX.

rank=128, alpha=default/unspecified, lr=1e-4, optimizer=adafactor

I'm training on a 4080 so it's very slow, not sure if I should stop to try a run without reg images and lose all progress if that's not the issue, guess I'll see if anyone else has tried it

**Comment by Tophness** [comment]
> Edit: Please try training without regularization images. I am not sure how the reg images work on FLUX.

I stopped the last run with reg images at 11 epochs. Loss was ~0.435 every epoch.
Tried without reg images for 10 epochs at rank 16. Still hasn't even started trying to learn the concept. Loss is 0.444-0.446 every epoch.

**Comment by sipie800** [comment]
> If you use Adafactor optimizer, the option `--optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False"` is needed for the training less than 10,000-20,000 steps. AdamW8bit may be worth to try.
> 
> In addition, please increase the strength to around 2.0 in generating images.

Hi, I can train flux lora with AdamW8bit. However with AdamW, the loss goes NAN at just 2nd step. Try many tuning it's always the same. Just for curiousity, any tips to train with a simple AdamW ? Or it's a known issue?

**Comment by Tophness** [comment]
> If you use Adafactor optimizer, the option `--optimizer_args "relative_step=False" "scale_parameter=False" "warmup_init=False"` is needed for the training less than 10,000-20,000 steps. AdamW8bit may be worth to try.

Have been using those optimizer_args. Was at 7,838 steps with reg images, 4,063 now without.
Adafactor may not be strong enough to do anything like a 326 image concept. I'll see how AdamW8bit goes on my 4080

**Comment by Tophness** [comment]
It was the reg images. Went from almost nothing to nearly fully trained around 9k steps.
[FurkanGozukara](https://github.com/kohya-ss/sd-scripts/issues?q=is%3Aissue+author%3AFurkanGozukara) confirmed reg images are counter-productive in a recent training video too.

**Comment by phyllispeng123** [comment]
> The default parameter is not proper. Need to lift the learning, I increase epoch from default 4 to 50 for real person with 50 images. Got pretty good result. It's better than SDXL. And the default lora dim is not proper IMO. I use 32 and 16. Flux is so good at details. To make good use of that, we may need higher rank in lora. And also, we may need higher quality with the images. If you use images with some artifects, the details (like hair texture...) may go bad than the origin flux. That would be a big mistake if you are using an excellent base model. To make good lora, we should take more care of the samples than with SDXL etc.

@sipie800 Can you kindly share your training script?  I meet the same issue now that there is no clear difference between the lora and the original model
